gateway/gateway.go

```diff
@@ -185,18 +200,15 @@ func (g *Gateway) Close() error {
 // Reconnect tries to reconnect forever. It will resume the connection if
 // possible. If an Invalid Session is received, it will start a fresh one.
 func (g *Gateway) Reconnect() {
-	for {
-		if err := g.ReconnectCtx(context.Background()); err != nil {
-			g.ErrorLog(err)
-		} else {
-			return
-		}
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), g.WSTimeout)
+	defer cancel()
+
+	g.ReconnectCtx(ctx)
 }
 
 // ReconnectCtx attempts to reconnect until context expires. If context cannot
 // expire, then the gateway will try to reconnect forever.
-func (g *Gateway) ReconnectCtx(ctx context.Context) (err error) {
+func (g *Gateway) ReconnectCtx(ctx context.Context) {
```

Comment:
ReconnectCtx should still return an error, I think. Having something that takes in a context and may timeout not returning an error is kind of counter-intuitive.

Perhaps have the `FatalErrorCallback` ran before returning err.